src/transformers/models/cogvlm/modeling_cogvlm.py

```diff
@@ -0,0 +1,1132 @@
+
+        mixed_raw_layer = torch.empty(
+            (batch_size, q_len, hidden_size * 3), dtype=hidden_states.dtype, device=hidden_states.device
+        )
+        mixed_raw_layer[vision_token_mask] = self.vision_expert_query_key_value(hidden_states[vision_token_mask])
+        mixed_raw_layer[language_token_mask] = self.language_expert_query_key_value(hidden_states[language_token_mask])
+
+        query_states, key_states, value_states = torch.split(mixed_raw_layer, self.hidden_size, dim=-1)
+        query_states = self._transpose_for_scores(query_states)
+        key_states = self._transpose_for_scores(key_states)
+        value_states = self._transpose_for_scores(value_states)
+
+        kv_seq_len = key_states.shape[-2]
+        if past_key_value is not None:
+            kv_seq_len += past_key_value[0].shape[-2]
+
+        cos, sin = self.rotary_emb(value_states, seq_len=position_ids.max() + 1)
+        cos, sin = (
+            nn.functional.embedding(position_ids, cos.squeeze(1)),
+            nn.functional.embedding(position_ids, sin.squeeze(1)),
+        )
+        query_states, key_states = apply_rotary_pos_emb(
+            query_states, key_states, cos, sin, position_ids, unsqueeze_dim=1
+        )
+
+        if past_key_value is not None:
+            key_states = torch.cat([past_key_value[0], key_states], dim=2)
+            value_states = torch.cat([past_key_value[1], value_states], dim=2)
+
+        past_key_value = (key_states, value_states) if use_cache else None
+
+        context_layer = torch.nn.functional.scaled_dot_product_attention(
+            query_states,
+            key_states,
+            value_states,
+            attn_mask=None,
+            dropout_p=0.0,
+            is_causal=True,
+        )
+
+        if context_layer.size() != (batch_size, self.num_heads, q_len, self.head_dim):
+            raise ValueError(
+                f"`attn_output` should be of size {(batch_size, self.num_heads, q_len, self.head_dim)}, but is"
+                f" {context_layer.size()}"
+            )
+        context_layer = context_layer.transpose(1, 2).contiguous().reshape(batch_size, q_len, self.hidden_size)
+
+        attn_output = torch.empty(context_layer.shape, dtype=hidden_states.dtype, device=hidden_states.device)
+        attn_output[vision_token_mask] = self.vision_expert_dense(context_layer[vision_token_mask])
+        attn_output[language_token_mask] = self.language_expert_dense(context_layer[language_token_mask])
+
+        return (attn_output, None, past_key_value)
+
+
+COGVLM_ATTENTION_CLASSES = {
+    "eager": CogvlmVisionExpertAttention,
+    "sdpa": CogvlmVisionExpertSdpaAttention,
+}
+
+
+class CogvlmDecoderLayer(nn.Module):
+    def __init__(self, config):
+        super().__init__()
+        self.hidden_size = config.hidden_size
+        # TODO eager isn't giving logits with same tolerance
+        config._attn_implementation = "sdpa"
```

Comment:
And another question for @fxmarty:

* currently only `attn_implementation="sdpa"` (in other words `CogvlmVisionExpertSdpaAttention`) gives equivalent logits with the original implementation.
* the "eager" implementation (`CogvlmVisionExpertAttention`) results in different logits (not passing even with atol=1e-1. The function `_prepare_4d_causal_attention_mask` is used to prepare the attention mask for the model.

Any idea what could be the reason? I'm pretty sure this has to do with the `attention_mask`

Reply:
Gently reminder here @fxmarty since this is the only thing left before this model can be merged.